### PR TITLE
Add client.forms.update to publish form updates

### LINKED ISSUE #6: Add client.forms.update to publish form updates (form def + optional attachments and attachments only)
## Use cases
Some folks make periodic updates to their form definitions on disk and/or in a shared network resource (e.g Google Drive) and would like their published forms on Central to always reflect those changes.

Some folks have entities in CSV/XML/GeoJSON attachments that they need to periodically update based on form submissions, updates to an external system, etc. In that case they want to update their entity lists but keep the same form definition.

Sample script: https://gist.github.com/lognaturel/2b251e54a3f3fe1e233435019b15ee17

### Variant 1 to update the form definition and optionally some attachments:
`form_id: str`
`xlsform: str` - path to XLSForm

`project_id: Optional[int] = None`
`attachments: Optional[List[str]] = None` - list of paths to files to attach to the form

### Variant 2 to update attachments and keep existing form definition:
`form_id: str`
`attachments: List[str]` - list of paths to files to attach to the form

`version_updater: Optional[Callable[string, string]]` - if specified, the currently-published form version is fetched, it's passed into the function, and the result is used as the version for publish. If not specified, `datetime.now().isoformat()` is used as the new version.
`project_id: Optional[int] = None`

## Proposed outputs
Exception in case anything goes wrong at any point -- XLSForm validation error, draft publishing error, etc.

I don't think any return value is necessary for the first variant. For the second variant, at least the version should be returned. In both cases it would be ok to return some kind of object representing form metadata though I think it would generally be ignored (except maybe for the version).

### PR DESCRIPTION
**Pull request by lindsay-stevens**
Closes #6

Branched from `pyodk-8` assuming that #19 is more or less good to go.

- forms.py:
  - new public method "update" for an existing form
  - helpers:
    - form_drafts.py for create / publish
    - form_draft_attachments.py: for upload
- validators.py: add validator for bool and path
- test_client.py: add integration tests, since the new helper methods either error out or return {"success": "True"}

#### What has been done to verify that this works as intended?

Added integration tests to check it works against staging. As mentioned above the new endpoints generally just return a bool so the mocks wouldn't have been very useful. Could add some mocked tests to check the internal logic, like setting the content_type header, or calling the version_updater, etc. if desired.

#### Why is this the best possible solution? Were any other approaches considered?

Not really - putting all the new code in just one function would've made it a bit too long to understand.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Shiny new feature!

#### Do we need any specific form for testing your changes? If so, please attach one.

The integration tests use forms that are saved in `tests/resources/forms`. So you'd just need a project and a pyodk config file, and upload a copy of the forms. The only difference between "range" and "range_draft" is the name in `instance/data/@id`.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.

Yes, the README is updated.

#### Before submitting this PR, please make sure you have:
- [x] included test cases for core behavior and edge cases in `tests`
- [x] run `nosetests` and verified all tests pass
- [x] run `python bin/pre_commit.py` to format / lint code
- [x] verified that any code or assets from external sources are properly credited in comments

### CODE REVIEW

**Comment by lindsay-stevens** [comment]
Rebased onto master.

**Review by lognaturel — CHANGES_REQUESTED** [review]
Shoot, I had this comment queued and didn't send. :/ Should be a quick change.

File: pyodk/_endpoints/forms.py [hunk 1]
```diff
@@ -95,3 +103,47 @@ def get(
...
             return Form(**data)
+
+    def create_version(
+        self,
+        form_id: str,
+        project_id: Optional[int] = None,
+        definition: Optional[str] = None,
+        attachments: Optional[Sequence[str]] = None,
+        version_updater: Optional[Callable[[str], str]] = None,
+    ) -> str:
```
[the remark this answers is not in the document]
Comment (lindsay-stevens) [reply pyodk/_endpoints/forms.py:114]: There seemed to be a large overlap in use case and functionality, and a general goal to keep the API small, so it seemed OK to combine the proposed variants. Could you please let me know what names you would like these functions to have, and how to describe them in the docstring / readme? Alternatively, I can change the publish call to not set a version when there's a new form definition. In either case I probably should add tests for the attachment-only use case. I think it's pretty useful for it to succeed in creating a new version no matter what combo of optional kwargs is provided.

[the remark this answers is not in the document]
Comment (lognaturel) [reply pyodk/_endpoints/forms.py:114]: I see two distinct workflows:
1. Update a form definition as it's being manually worked on. I don't think this will be super common but it's a convenience alternative to updating through the frontend. In that case, the `version` is managed in the form definition and the user should be updating it.
1. Script a form attachment update. This is useful for synchronizing with an external data source or updating entities based on form submissions that have come in. In that case the form version generally doesn't matter (the definition isn't changing anyway).

I'd be ok with only adding the second case above for now.

With two methods, I would call them both `update` and use the following descriptions:
- Update the form definition and optionally its attachment(s). The form definition MUST specify a new `version`.
- Update the form's attachment(s). By default, the form version is updated with a timestamp at time of publish. You can also specify a reference to a function that will be applied to the current form's version. It should return the next intended version (by e.g. adding 1 or changing a suffix).

Here are the cases that a single method would need to address:
- form definition with new version, no version_updater: that should work and the version from the def should be used
- form definition with existing version, no version_updater: backend will error and that error should be passed on (should NOT default to timestamp here I don't think because it could mask an error)
- form definition with new version, version_updater: this is weird and is the big thing I was trying to avoid with the two variants. Detecting the case would require parsing the form definition which we don't want to do. So I think our best option is to ignore the version_updater in that case. 
- form definition with existing version, version_updater: I did not intend this to be supported but I think it's ok. We can detect a [`409.6`](https://github.com/getodk/central-backend/blob/467fc0b81b6a48b794194eb724e5d9d4c07971e7/lib/util/problem.js#L160), get the current version, apply `version_updater`, retry with the new version

I find it easier to separate out the two workflows but as long as those cases are handled reasonably and are tested, you can make the final call.

[the remark this answers is not in the document]
Comment (lindsay-stevens) [reply pyodk/_endpoints/forms.py:114]: Thanks! This extra detail is helpful.

Python doesn't really have multiple dispatch or method overloading like Java or some other languages. Everything is a dict so (if not doing anything special) with duplicate names the last defined method wins. These features can be simulated with helper libraries, but I think it's an unusual pattern for Python. There is single dispatch in the `functools` library, but then the first args would need to be different e.g. `update(def, xid, attach)` and `update(attach, xid)` and I'm not sure if they can have kwargs or different optional args. Similarly there's `typing.overload` but the implementation ends up in one function with a need to check args, i.e. it's mainly an annotation thing.

So if you would like a single method I'd go for the existing catch-all and making sure it's tested against these use cases. Or as separate methods, maybe they could be called `update_definition` and `update_attachments`?

[the remark this answers is not in the document]
Comment (lognaturel) [reply pyodk/_endpoints/forms.py:114]: Or how about we omit `version_updater` for now. If a form definition is provided, we enforce that it specify a new version and error otherwise. If a form definition is not provided, we always use a timestamp version. That makes the logic simpler for now and it's probably enough for nearly all cases.
